File: backend/schemas/company.py
from typing import List

from pydantic import BaseModel


class CompanySymbol(BaseModel):
    symbol: str
    latest_price: float | None = None
    latest_date: str | None = None

    model_config = {"from_attributes": True}


class CompanyListResponse(BaseModel):
    companies: List[CompanySymbol]
    total: int


class MarketCapBucket(BaseModel):
    label: str        # e.g. "100–1K Cr"
    category: str     # e.g. "Micro Cap"
    min_cr: float
    max_cr: float | None   # None = open-ended (last bucket)
    count: int
    total_cap_cr: float


class MarketStatsResponse(BaseModel):
    total_symbols: int
    total_market_cap_cr: float
    latest_date: str | None
    buckets: List[MarketCapBucket]


class MarketCapTrendPoint(BaseModel):
    """One data point in the market cap time series."""
    date: str
    total_market_cap_cr: float
    total_companies: int
    nano_count: int
    micro_count: int
    small_count: int
    mid_count: int
    large_count: int


class MarketCapTrendResponse(BaseModel):
    """Time series of market cap and company count by bucket."""
    interval: str
    start_date: str
    end_date: str
    total_points: int
    data: List[MarketCapTrendPoint]

File: frontend-next/src/types/company.ts
export interface CompanySymbol {
  symbol: string;
  latest_price?: number;
  latest_date?: string;
}

export interface CompanyListResponse {
  companies: CompanySymbol[];
  total: number;
}

export interface MarketCapBucket {
  label: string;
  category: string;
  min_cr: number;
  max_cr: number | null;
  count: number;
  total_cap_cr: number;
}

export interface MarketStatsResponse {
  total_symbols: number;
  total_market_cap_cr: number;
  latest_date: string | null;
  buckets: MarketCapBucket[];
}

export interface MarketCapTrendPoint {
  date: string;
  total_market_cap_cr: number;
  total_companies: number;
  nano_count: number;
  micro_count: number;
  small_count: number;
  mid_count: number;
  large_count: number;
}

export interface MarketCapTrendResponse {
  interval: string;
  start_date: string;
  end_date: string;
  total_points: number;
  data: MarketCapTrendPoint[];
}
